lalrpop/src/build/mod.rs

```diff
@@ -205,42 +207,29 @@ fn needs_rebuild(lalrpop_file: &Path, rs_file: &Path) -> io::Result<bool> {
 
 fn lalrpop_files<P: AsRef<Path>>(root_dir: P) -> io::Result<Vec<PathBuf>> {
     let mut result = vec![];
-    for entry in fs::read_dir(root_dir)? {
+
+    let walkdir = WalkDir::new(root_dir)
+        .follow_links(true)
+        // Use deterministic ordering:
+        .sort_by_file_name()
+        // Prevent loops:
+        .same_file_system(true);
+    for entry in walkdir {
         let entry = entry?;
-        let file_type = entry.file_type()?;
+        // `file_type` follows symlinks, so if `entry` points to a symlink to a file, then
+        // `is_file` returns true.
+        if !entry.file_type().is_file() {
+            continue;
+        }
 
         let path = entry.path();
-
-        if path.is_dir() {
-            result.extend(lalrpop_files(&path)?);
+        if path.extension() != Some(OsStr::new("lalrpop")) {
```

Comment:
It's kind of a micro-optimization to ask if this check could have been implemented with https://docs.rs/walkdir/latest/walkdir/struct.IntoIter.html#method.filter_entry . More of a thought than any thing else, might not be worth changing.

Reply:
Yes, I saw this function as well. However, it looks like `filter_entry` is there mostly to skip traversing into subdirectories, not to skip individual files. We don't want to skip directories, so the lambda we pass to `filter_entry` would need to return `true` for directories. But then we'd consume the (now filtered) `WalkDir` iterator, we'd need to skip over directories. I think the current approach is simpler and not less efficient than an approach using `filter_entry`.

Reply:
I was thinking about this because the files in a directory are sorted first, then we ignore everything that isn't a lalrpop file. It's probably not a bottleneck and as you say, probably not worth it over the simpler solution.

Reply:
Ah, right, if the sorting happens after filtering, then a solution using `filter_entry` might actually be faster. I don't think we'll notice a difference in almost all cases though.